seb_openedx/middleware.py

```diff
@@ -1,27 +1,132 @@
+            access_denied = True
+
+            if self.is_whitelisted_view(request, course_key):
+                # First: Broad white-listing
+                access_denied = False
+
+            if self.is_blacklisted_chapter(request, course_key):
+                # Second: Granular white-listing
+                access_denied = True
+
             active_comps = get_enabled_permission_classes()
             for permission in active_comps:
                 if permission().check(request, course_key):
-                    return
-            return render_to_response('seb-403.html', status=403)
+                    access_denied = False
+
+        if access_denied:
+            is_courseware_view = bool(view_func.__name__ == courseware.views.index.CoursewareIndex.__name__)
+            if is_courseware_view:
+                return self.courseware_error_response(request, *view_args, **view_kwargs)
+            return self.generic_error_response(request, course_key)
+        return None
+
+    def is_whitelisted_view(self, request, course_key):
+        """ First broad filter: whitelisting of paths/tabs """
+        views_module = inspect.getmodule(request.resolver_match.func).__name__
+
+        if 'discussion' in SEB_WHITELIST_PATHS and views_module == 'discussion.views':
+            return True
+
+        if 'wiki' in SEB_WHITELIST_PATHS and views_module == 'course_wiki.views':
+            return True
+
+        if 'courseware' in SEB_WHITELIST_PATHS and self.is_xblock_request(request):
+            return True
+
+        url_names_allowed = list(SEB_WHITELIST_PATHS)
+
+        if 'course-outline' in SEB_WHITELIST_PATHS:
+            url_names_allowed.append('openedx.course_experience')
+
+        if request.resolver_match.url_name:
+            if request.resolver_match.url_name in ['jump_to', 'jump_to_id']:
+                # Allow redirects:
+                return True
+            for url_name in url_names_allowed:
+                if request.resolver_match.url_name.startswith(url_name):
+                    return True
+        return False
+
+    def is_blacklisted_chapter(self, request, course_key):
+        """ Second more granular filter: blacklisting of specific chapters """
+        chapter = request.resolver_match.kwargs.get('chapter')
+
+        if not SEB_BLACKLIST_CHAPTERS:
+            return False
+
+        if chapter in SEB_BLACKLIST_CHAPTERS:
+            return True
+
+        if 'courseware' in SEB_WHITELIST_PATHS and self.is_xblock_request(request):
+            modulestore = get_xmodule_modulestore_module().django.modulestore
+            path_to_location = get_xmodule_modulestore_module().search.path_to_location
+            usage_id = request.resolver_match.kwargs.get('usage_id')
+            usage_key = UsageKey.from_string(unquote_slashes(usage_id)).map_into_course(course_key)
+            if usage_key:
+                path = path_to_location(modulestore(), usage_key)
+                chapter = path[1]
```

Comment:
The 1 is acting as a Magic Number. Can we refactor that?

Reply:
I guess we can assign 1 to a variable called chapter_index and then use that variable (beyond that not much is possible without using reflection to check what argument returns the chapter)